loopy/statistics.py

```diff
@@ -643,43 +717,59 @@ class Op(ImmutableRecord):
 
         A :class:`str` representing the kernel name where the operation occurred.
 
-    """
-
-    def __init__(self, dtype=None, name=None, count_granularity=None,
-            kernel_name=None):
-        if count_granularity not in [*CountGranularity.ALL, None]:
-            raise ValueError("Op.__init__: count_granularity '%s' is "
-                    "not allowed. count_granularity options: %s"
-                    % (count_granularity, [*CountGranularity.ALL, None]))
+    .. attribute:: tags
 
-        if dtype is not None:
+        A :class:`frozenset` of tags to the operation.
+    """
+    dtype: LoopyType | None = None
+    op_type: OpType | None = None
+    count_granularity: CountGranularity | None = None
+    kernel_name: str | None = None
+    tags: frozenset[Tag] = frozenset()
+
+    def __post_init__(self):
+        if self.dtype is not None:
             from loopy.types import to_loopy_type
-            dtype = to_loopy_type(dtype)
+            object.__setattr__(self, "dtype", to_loopy_type(self.dtype))
+
+        assert isinstance(self.op_type, (OpType, type(None)))
 
-        super().__init__(dtype=dtype, name=name,
-                        count_granularity=count_granularity,
-                        kernel_name=kernel_name)
+        if not isinstance(self.count_granularity, (CountGranularity, type(None))):
+            raise ValueError(
+                f"unexpected count_granularity: '{self.count_granularity}'")
 
     def __repr__(self):
-        # Record.__repr__ overridden for consistent ordering and conciseness
         if self.kernel_name is not None:
-            return (f"Op({self.dtype}, {self.name}, {self.count_granularity},"
-                    f' "{self.kernel_name}")')
+            return (f"Op({self.dtype}, {self.op_type}, {self.count_granularity},"
+                    f' "{self.kernel_name}", {self.tags})')
         else:
-            return f"Op({self.dtype}, {self.name}, {self.count_granularity})"
+            return f"Op({self.dtype}, {self.op_type}, " + \
+                        f"{self.count_granularity}, {self.tags})"
 
 # }}}
 
 
 # {{{ MemAccess descriptor
 
-class MemAccess(ImmutableRecord):
+class AccessDirection(Enum):
+    """
+    Specify the direction of a memory access.
+
+    .. attribute:: READ
+    .. attribute:: WRITE
+    """
+    READ = 0
+    WRITE = 1
+
+
+@dataclass(frozen=True, eq=True)
+class MemAccess:
     """A descriptor for a type of memory access.
```

Comment:
Move to autoattribute and in-line docstrings?

Reply:
Done in 6959904c